Makefile

```
@@ -1,37 +1,105 @@
-PYTHON ?= python
-POETRY ?= poetry
-PRE_COMMIT ?= pre-commit
-TOX ?= tox
+VENV_DIR ?= .venv
+
+SYSTEM_PYTHON ?= python
+
+# If poetry already exists, use it.  If not, we'll create a local venv for it
+POETRY ?= $(or $(shell command -v poetry 2>/dev/null),$(VENV_DIR)/bin/poetry)
+# Calculate it once
+POETRY := $(POETRY)
+MYPY ?= $(POETRY) run mypy
+PYTHON ?= $(POETRY) run python
+PRE_COMMIT ?= $(POETRY) run pre-commit
+TOX ?= $(POETRY) run tox
+
+# Note: any comment that starts with '## ' is taken to be a help string and
+# emitted by the 'help' target, along with the line above it with a leading
+# ".PHONY: " or ".HELP: " removed.  Any other commentary must use just one '#'.
+
+.PHONY: help
+## Display this help message
+help: _ESCAPE := $(shell printf '\033')
+help:
+	@echo "Usage: make [target]"
+	@grep --no-group-separator -EB1 '^## ' Makefile \
+	| sed \
+	-e 's/^.\(PHONY\|HELP\): \(.\+\)/\n$(_ESCAPE)[36m\2$(_ESCAPE)[0m:/' \
+	-e 's/^## /\t/'
+	@echo
+	@echo "If available, a global installation of \`poetry\` will be used \
+	       to manage this project.  Otherwise, a local virtual \
+	       environment will be created in $(VENV_DIR) using \
+	       \`$(SYSTEM_PYTHON)\`.  Other commands such as \`mypy\` or \
+	       \`tox\` will be run using \`poetry run\`.  There are several \
+	       makefile variables set at the top of Makefile that may be \
+	       defined in your environment or on the command line to affect \
+	       how various targets are run." \
+	       | sed -e 's/\s\s\+/ /g' \
+	       | fmt -w 80
 
 .PHONY: all
-all: data poetry.lock lint test
+## Run all commands necessary to ensure that everything is up to date and
+## passing existing tests, then build distribution artifacts
+all: update data check
 	$(POETRY) build
 
 .PHONY: data
-data:
+## Regenerate all ISO data from the upstream Debian iso-codes project
+data: poetry.lock
 	rm -rf src/pycountry/databases
 	rm -rf src/pycountry/locales
 	$(PYTHON) generate.py
 
 .PHONY: sdist
-sdist: data poetry.lock
+## Create a source distribution of the project
+sdist: data
 	$(POETRY) build --format=sdist
 
 .PHONY: wheel
-wheel: data poetry.lock
+## Create a wheel distribution of the project
+wheel: data
 	$(POETRY) build --format=wheel
 
-poetry.lock: pyproject.toml
-	$(POETRY) lock
+.HELP: poetry.lock
+## Ensure that the poetry.lock file is up to date and install all dependencies
+## in the poetry environment
+poetry.lock: pyproject.toml | $(POETRY)
```

I'm not familiar with this syntax, and when removing the venv it doesn't seem like this is installed :thinking: , is this what is meant?

```suggestion
poetry.lock: pyproject.toml $(POETRY)
```

It's now always running these commands, but they seem to be pretty fast

---

The `|` notes it as an ["ordering only prerequisite"](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html), meaning (IIUC, makefiles are arcane...) that it basically just checks that the file exists, but doesn't force a regeneration of *this* file if that one has changed.